Trac 53489: Set sidebar state when rendering widgets block editor screen

hellofromtonya opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/53489

Uses the same strategy as the classic widgets screen by running `retrieve_widgets()`. In doing so, newly added sidebars are displayed in the Widgets Block Editor.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

hellofromtonya committed:
Adds test for newly registered sidebar.

hellofromtonya committed:
Sets initial sidebars and widgets state when getting items.

hellofromtonya committed:
Updates tests to reset global and init widgets.

noisysocks commented on tests/phpunit/tests/rest-api/rest-sidebars-controller.php:
```diff
@@ -128,6 +141,8 @@ public function test_context_param() {
 	 * @ticket 41683
 	 */
 	public function test_get_items() {
+		wp_widgets_init();
```
Should this go in `setUp()` so that we can avoid some repetition?

hellofromtonya replied:
Explored doing that. But it skews all the other tests that don't have it. In reviewing other widget tests, this init function is placed in each test that needs it. Not sure why yet. But it's on my test review TODO list across the test suite.

TimothyBJacobs commented:
We probably need to add `retrieve_widgets` to the other callbacks too, no?

hellofromtonya commented:
>We probably need to add retrieve_widgets to the other callbacks too, no?

@TimothyBJacobs such as? I'm seeing the problem anywhere else. It's updating, removing, and rendering as expected. Do you foresee another problem area that could use it?

TimothyBJacobs commented:
I'm not sure to what degree Gutenberg is using those other callbacks and their responses. But if the data is missing for the `get_items` call, then won't it also be missing when getting a single sidebar?

hellofromtonya commented:
> I'm not sure to what degree Gutenberg is using those other callbacks and their responses. But if the data is missing for the `get_items` call, then won't it also be missing when getting a single sidebar?

Hmm, that's a good point. I'm not seeing that effect. But let me do a little bit more testing to see what happens with `get_item` callback.

hellofromtonya committed:
Set initial set in `get_item()` too.

hellofromtonya commented:
>I'm not sure to what degree Gutenberg is using those other callbacks and their responses. But if the data is missing for the get_items call, then won't it also be missing when getting a single sidebar?

@TimothyBJacobs In testing, seems `get_item` is _currently_ not called in `Widgets` or `Customizer` or in the frontend. 

What if it were called in the future or by another plugin/theme? 
Added it and did more testing. Also ran the test suite which passed. I think you're right. Added it to `get_item()` in this PR to guard against _future_ or _yet to be reported_ issues. (See commit https://github.com/WordPress/wordpress-develop/pull/1422/commits/ebef519b867f1cf82980b4a2efd0e492a83a4e82)

TimothyBJacobs commented:
Great, let's ship it!

TimothyBJacobs commented:
Merged in https://core.trac.wordpress.org/changeset/51235.
